File: array_c_plus_plus/34.cpp
#include<vector>
#include<iostream>

using namespace std; 

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> result(2);
        result[0] = -1;
        result[1] = -1;

        int low=0, high=nums.size()-1, mid;
        
        while(high >= low){
            // cout<<mid<<endl;
            mid = low + (high - low) / 2;
            if(nums[mid] > target){
                high = mid - 1;
            }else if(nums[mid] < target){
                low = mid + 1;
            }else{
                // cout<<mid<<endl;
                int i=mid, j=mid;
                while(nums[i]==target && i>-1)i--;
                // cout<<i<<endl;
                i++;
                while(nums[j]==target && j<nums.size()) j++;
                j--;
                result[0] = i;
                result[1] = j;
                break;
            }
        }
        return result;
    }
};


int main(){
    int a[] = {5, 7, 7, 8, 8, 10};
    vector<int> v(a, a+6);
    
    Solution s;
    vector<int> result = s.searchRange(v, 8);
    cout<<result[0]<<" "<<result[1]<<endl;
    return -1;
}

File: array_python/34.py
# coding=UTF-8
'''
@Author: httermin
@Date: 2020-01-01 17:15:24
'''
import sys
import os

sys.path.append("..")
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        
        start, end = 0, len(nums) - 1
        result_idx = -1
        while start <= end:
            mid = (start + end) // 2
            if nums[mid] == target:
                result_idx = mid
                break

            if nums[mid] > target:
                end = mid - 1
            else:
                start = mid + 1
        
        if result_idx == -1:
            return [-1, -1]
        
        lo, hi = mid, mid

        while lo > 0 and nums[lo-1] == target:
            lo -= 1
        
        while hi+1 < len(nums) and nums[hi+1] == target:
            hi += 1
        
        return [lo, hi]

if __name__ == "__main__":
    solu = Solution()

    example = [([1], 1), ([5,7,7,8,8,10], 8), ([5,7,7,8,8,10], 6), ([5,7,8,8,8,10], 8), ([5,7,7,8,8,8,10], 8)]

    for ex, t in example:
        print(solu.searchRange(ex, t))